sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,278 @@
+/*!
+# Introduction
+
+shibaken is called by sundog as a setting generator.
+
+shibaken will fetch and populate the host container's user-data with authorized ssh keys from the
+AWS instance metadata service (IMDS).
+
+(The name "shibaken" comes from the fact that Shiba are small, but agile, hunting dogs.)
+*/
+
+#![deny(rust_2018_idioms)]
+
+#[macro_use]
+extern crate log;
+
+use reqwest::blocking::Client;
+use serde::Serialize;
+use simplelog::{Config as LogConfig, LevelFilter, TermLogger, TerminalMode};
+use snafu::{ensure, ResultExt};
+use std::str::FromStr;
+use std::{env, process};
+
+// Instance Meta Data Service.
+//
+// Currently only able to get fetch session tokens from `latest`
+// FIXME Pin to a date version that supports IMDSv2 once such a date version is available.
+const IMDS_PUBLIC_KEY_BASE_URL: &str = "http://169.254.169.254/latest/meta-data/public-keys/";
+const IMDS_SESSION_TOKEN_ENDPOINT: &str = "http://169.254.169.254/latest/api/token";
+
+#[derive(Serialize)]
+struct UserData {
+    ssh: Ssh,
+}
+
+#[derive(Serialize)]
+struct Ssh {
+    authorized_keys: Vec<String>,
+}
+impl UserData {
+    fn new(public_keys: Vec<String>) -> Self {
+        UserData {
+            ssh: Ssh {
+                authorized_keys: public_keys,
+            },
+        }
+    }
+}
+
+/// Helper to fetch an IMDSv2 session token that is valid for 60 seconds.
+fn fetch_imds_session_token(client: &Client) -> Result<String> {
+    let uri = IMDS_SESSION_TOKEN_ENDPOINT;
+    let imds_session_token = client
+        .put(uri)
+        .header("X-aws-ec2-metadata-token-ttl-seconds", "60")
+        .send()
+        .context(error::ImdsRequest { method: "PUT", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })?;
+    Ok(imds_session_token)
+}
+
+/// Helper to fetch data from IMDS.
+fn fetch_data_from_imds(client: &Client, uri: &str, session_token: &str) -> Result<String> {
+    client
+        .get(uri)
+        .header("X-aws-ec2-metadata-token", session_token)
+        .send()
+        .context(error::ImdsRequest {
+            method: "GET",
+            uri: uri.to_string(),
+        })?
+        .error_for_status()
+        .context(error::ImdsResponse {
+            uri: uri.to_string(),
+        })?
+        .text()
+        .context(error::ImdsText {
+            uri: uri.to_string(),
+        })
+}
+
+/// Returns a list of public keys.
+fn fetch_public_keys_from_imds() -> Result<Vec<String>> {
+    info!("Fetching IMDS session token");
+    let client = Client::new();
+    let imds_session_token = fetch_imds_session_token(&client)?;
+
+    info!("Fetching list of available public keys from IMDS");
+    // Returns a list of available public keys as '0=my-public-key'
+    let public_key_list =
+        fetch_data_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
```

Comment:
My misgiving here is that we're essentially parsing an index page. It might be cleaner and simpler just to start from `0` and retrieve the data until we get a 404.

Reply:
That is certainly a valid approach. I would rather keep the current logic as it more closely reflects the logic in the [admin container's entrypoint](https://github.com/bottlerocket-os/bottlerocket-admin-container/blob/eec7430012182bb77a68a327f10065abdcc73a3f/start_admin_sshd.sh#L21) and the [examples for retrieving public keys](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instance-metadata-ex-3) in the AWS documentation.

Reply:
Fair enough!